tests/test_tomography_gate_set_tomography.py

```diff
@@ -0,0 +1,2535 @@
+
+    inverse_circuit = reset_register(test_circuit, "sp_t")
+
+    assert isinstance(inverse_circuit, qibo.models.circuit.Circuit)
+
+
+def test_reset_register_invalid_string():
+    # Test resetting both qubits
+
+    nqubits = 2
+    test_circuit = qibo.models.Circuit(nqubits)
+    test_circuit.add(gates.H(0))
+    test_circuit.add(gates.CNOT(0, 1))
+
+    # Check if NameError is raised
+    with pytest.raises(NameError):
+        inverse_circuit = reset_register(test_circuit, "sp_2")
+
+
+#################################################################################
+
+
+def test_GST_execute_circuit_1qb_j0():
+    np.random.seed(42)
+    nqubits = 1
+    circuit = qibo.models.Circuit(nqubits)
+    circuit.add(gates.M(0))
+    k = 0
+    j = 0
+    result = GST_execute_circuit(circuit, k, j)
+    assert result == 1.0
+
+
+def test_GST_execute_circuit_1qb_jnonzero():
+    np.random.seed(42)
+    nqubits = 1
+    circuit = qibo.models.Circuit(nqubits)
+    circuit.add(gates.H(0))
+    circuit.add(gates.M(0))
+    k = 0
+    j = 2
+    result = GST_execute_circuit(circuit, k, j)
+    seed42_value = 0.015200000000000047
+    assert result == pytest.approx(seed42_value, abs=1e-12)
+
+
+def test_GST_execute_circuit_2qb_j0():
+    np.random.seed(42)
+    nqubits = 2
+    circuit = qibo.models.Circuit(nqubits)
+    circuit.add(gates.H(0))
+    circuit.add(gates.M(0))
+    circuit.add(gates.M(1))
+    k = 0
+    j = 0
+    result = GST_execute_circuit(circuit, k, j)
+    assert result == 1.0
+
+
+def test_GST_execute_circuit_2qb_jnonzero():
+    np.random.seed(42)
+    nqubits = 2
+    circuit = qibo.models.Circuit(nqubits)
+    circuit.add(gates.H(0))
+    circuit.add(gates.H(1))
+    circuit.add(gates.M(0))
+    circuit.add(gates.M(1))
+    k = 0
+    j = 4
+    result = GST_execute_circuit(circuit, k, j)
+    seed42_value = 0.015200000000000075
+    assert result == pytest.approx(seed42_value, abs=1e-12)
+
+
+def test_GST_execute_circuit_wrong_qb():
+    nqubits = 4
+    circuit = qibo.models.Circuit(nqubits)
+    circuit.add(gates.M(0))
+    k = 0
+    j = 0
+
+    # Check if ValueError is raised
+    with pytest.raises(ValueError, match="nqubits needs to be either 1 or 2"):
+        result = GST_execute_circuit(circuit, k, j)
+
+
+#################################################################################
+
+
+def test_GST_one_qubit_empty_circuit():
+    np.random.seed(42)
+    nqubits = 1
+    result = execute_GST(
+        nqubits, gate=None, invert_register=None, noise_model=None, backend=None
+    )
+    seed42_result = np.array(
+        [
+            [1, 1, 1, 1],
+            [0.0152, 0, 1, 0.0052],
+            [-0.0128, 0.0126, 0.0018, 1],
+            [1, -1, -0.0114, -0.0034],
+        ]
+    )
+    assert result == pytest.approx(seed42_result, abs=1e-12)
+
+
+def test_GST_two_qubit_empty_circuit():
+    np.random.seed(42)
+    nqubits = 2
+    result = execute_GST(nqubits)
+    seed42_result = np.array(
```

Comment:
How did you get this? 
You should code the procedure used to get it rather than hard-code the result.
The same applies for the other instances below.

Reply:
Let's take ```nqubits=1```. For a circuit without a ```gate```, there will be 4 initial states and 4 measurement bases. Using the default ```nshots=int(1e4)``` for each of the 16 circuits, we get some distribution and compute the expectation value. A 4x4 matrix is used to store each of the 16 results. 

Please advice how to code the procedure rather check the hard coded results.

Reply:
I was not referring to the (4,4) matrix above, but to the (16,16) matrix below and the others following. Having a (4,4) matrix as target is acceptable, but we cannot afford to have 2000+ lines of code in this test just to store all the matrices. Since I doubt that you calculated by hand all the elements of the matrices, I assume you wrote the code to calculate those. What I am saying is: regenerate the target matrix at runtime rather than explicitly writing the target.

Reply:
I totally agree that it's quite ridiculous to have 2000+ lines for the 16x16 matrices, I didn't know a better way. 
I was thinking of fixing the random seed and store that outcome. Then I use pytest to check if the tests generate the same outcome.

In this case, can we generate two matrices at runtime then use ```np.allclose(matrix1, matrix2)``` as a check?

Reply:
Yes please, generate them at runtime and use the `backend.assert_allclose` method to check they are equal, take a look to the other tests in qibo. Try to use the pytest parameters and `backend` fixture to make the tests run on all the available backends as @renatomello was saying above.

Reply:
Sure thing, I'll incorporate the ```pytest.mark.parametrize``` too.